File: src/player/engine.rs
//! The platform-neutral playback engine seam.
//!
//! Port of `web/player-core/src/engine.ts` (BSD-3-Clause), reshaped from an
//! asynchronous callback port into a **synchronous** trait (Phase 10
//! deliberate deviation — the core has no async runtime). A host adapter that
//! wraps genuinely asynchronous platform engines drives this trait from its
//! own event loop and injects engine facts through the player's `on_*`
//! methods.
//!
//! Capability concepts from the TypeScript port are preserved:
//!
//! - `preload_next` gates the standby (`prepare_next`/`advance`) path;
//! - `crossfade` gates `begin_crossfade`;
//! - `decode_gate` is informational; the pump calls (`start_pumping` /
//!   `pause_pumping`) are always available and default to no-ops, exactly
//!   like the TypeScript `gate()` fallback.
//!
//! All positions are OUTPUT-rate samples within the currently open track.

use super::types::{PlaybackItem, StreamInfo};

/// Engine capability flags (`EngineCapabilities`).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct EngineCapabilities {
    /// Standby `prepare_next`/`advance` support.
    pub preload_next: bool,
    /// Track handoff is sample-exact.
    pub sample_accurate_gapless: bool,
    /// Decode-gate control is meaningful (pull-style pipelines).
    pub decode_gate: bool,
    /// Overlapped `begin_crossfade` transitions are supported.
    pub crossfade: bool,
}

/// Result of a taken crossfade transition (`CrossfadeResult`).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CrossfadeResult {
    /// The incoming track's stream facts.
    pub info: StreamInfo,
    /// Output-rate frames of overlap actually mixed.
    pub overlap_frames: u64,
}

/// Outcome of a `begin_crossfade` request.
///
/// TypeScript's `beginCrossfade` returns a `Promise` that may resolve
/// immediately or later. The synchronous seam represents that as three
/// cases: declined, completed now, or in flight (the host later calls
/// `Player::on_crossfade_complete`).
#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub enum CrossfadeStart {
    /// No fade: fall back to the normal boundary path.
    Declined,
    /// The fade completed synchronously.
    Completed(CrossfadeResult),
    /// The fade is in flight; the host will report completion.
    Pending,
}

/// An engine failure message (the TypeScript `error(message)` payload).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EngineError(pub String);

impl std::fmt::Display for EngineError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for EngineError {}

/// Engine operation result.
pub type EngineResult<T> = Result<T, EngineError>;

/// A host-implementable abstraction over "the thing that turns one or more
/// queued sources into audible, seekable output".
pub trait Engine {
    /// Declared capabilities (honest; nothing claims more than it does).
    fn capabilities(&self) -> EngineCapabilities;

    /// Opens a source; returns its stream facts.
    fn open(&mut self, item: &PlaybackItem) -> EngineResult<StreamInfo>;

    /// Starts/resumes output.
    fn play(&mut self) -> EngineResult<()>;

    /// Pauses output.
    fn pause(&mut self);

    /// Seeks within the OPEN track (output-rate samples).
    fn seek(&mut self, samples: u64);

    /// Sets the combined linear gain.
    fn set_gain(&mut self, linear: f64);

    /// Output-rate frames rendered since the last open/seek reset.
    fn rendered_samples(&self) -> u64;

    /// Releases the engine's resources.
    fn close(&mut self);

    /// Opens the next source ahead of time (standby). Default: unsupported.
    fn prepare_next(&mut self, _item: &PlaybackItem) -> Option<StreamInfo> {
        None
    }

    /// Promotes a previously prepared standby. `expected` is the current
    /// policy target; a standby prepared for anything else must be refused
    /// (`None`) without promoting. Default: unsupported.
    fn advance(&mut self, _expected: Option<&PlaybackItem>) -> Option<StreamInfo> {
        None
    }

    /// Begins an overlapped transition. Default: unsupported.
    fn begin_crossfade(&mut self, _next: &PlaybackItem, _fade_seconds: f64) -> CrossfadeStart {
        CrossfadeStart::Declined
    }

    /// True when the decoder is exhausted and the output has run dry.
    fn is_output_drained(&self) -> bool {
        false
    }

    /// Starts the decode pump (no-op when the engine has no gate).
    fn start_pumping(&mut self) {}

    /// Stops the decode pump (no-op when the engine has no gate).
    fn pause_pumping(&mut self) {}
}

File: web/player-core/src/engine.ts
// Copyright (c) 2026, The MusicPack Development Team
// SPDX-License-Identifier: BSD-3-Clause

// The playback engine port (player-core M2).
//
// A host-implementable abstraction over "the thing that turns one or more
// queued sources into audible, seekable output". The web hosts implement it
// with AudioContext/AudioWorklet/WASM (musepack) and HTMLMediaElement
// (native codecs); future Android/iOS hosts implement it natively. The core
// orchestrator knows nothing about audio plumbing beyond this contract.
//
// Honesty rules:
// - Capabilities declare what an engine actually does. Nothing here claims
//   sample-accurate gapless for element-based playback.
// - DecodeGate is a capability, not part of Engine: priming requires
//   decoding while the output clock is still stopped (a suspended
//   AudioContext pulls no render callbacks), so folding pump control into
//   play()/pause() would deadlock priming on pull-style pipelines.
// - All positions are OUTPUT-rate samples within the currently open track.
//   Source-rate conversion stays engine-internal.

import type { PlaybackItem, StreamInfo } from './types';

/** Backend families a host may select per item. `musepack` and `rust` are
 *  pull-decode demand-range engines whose reported position is session-relative
 *  (reset on open/seek, monotonic across gapless handoffs); `native` is
 *  element-backed and reports absolute within-track time. This is a host label
 *  only — the core learns nothing about a backend's internals. */
export type EngineKind = 'musepack' | 'native' | 'rust';

export interface EngineCapabilities {
  /** prepareNext()/advance() standby support. */
  readonly preloadNext: boolean;
  /** True only when track handoff is sample-exact (Musepack pipeline). */
  readonly sampleAccurateGapless: boolean;
  /** DecodeGate control is meaningful (pull-style decode pipelines). */
  readonly decodeGate: boolean;
  /** beginCrossfade() overlapped transitions are supported (M8). */
  readonly crossfade: boolean;
}

/** Events an engine reports back. Delivered asynchronously, in order,
 *  tagged with the engine instance (see plan §A.5 delivery rules). */
export type EngineEventName = 'primed' | 'buffering' | 'eos' | 'error' | 'tick';

export interface EngineEvents {
  /** Enough output is buffered to start/keep the clock running. */
  primed(): void;
  /** Output starved; playback cannot continue at full rate. */
  buffering(): void;
  /** The current source finished decoding/reached its end. */
  eos(): void;
  error(message: string): void;
  /** Coarse position change; may be coalesced by either side. */
  tick(): void;
}

export interface Engine {
  readonly capabilities: EngineCapabilities;

  init(authToken: string | null): Promise<void>;
  /** Opens a source; resolves once stream facts are known. */
  open(item: PlaybackItem): Promise<StreamInfo>;
  play(): Promise<void>;
  pause(): Promise<void>;
  /** Seeks within the OPEN track (output-rate samples). */
  seekSample(samples: number): Promise<void>;
  /** Combined linear gain (user volume × normalization). */
  setGain(linear: number): void;
  /** Output-rate frames rendered since the last open/seek reset. */
  renderedSamples(): number;
  close(): Promise<void>;

  on(name: EngineEventName, cb: (sender: Engine) => void): void;
}

/** Standby/preload capability: open the next source ahead of time and
 *  promote it at the boundary without re-opening the output graph.
 *
 *  Standby/policy agreement: `advance(expected)` receives the item the
 *  core's CURRENT queue policy selected (peeked at promotion time, not when
 *  the standby was prepared). The standby may be promoted only when it was
 *  prepared for exactly that item (`sameItemIdentity`); otherwise — or when
 *  `expected` is `null`, meaning nothing may follow — the engine returns
 *  `null` WITHOUT flushing or promoting, so a stale lane can never bleed
 *  into output or metadata. The caller recovers by loading `expected`
 *  fresh. */
export interface PreloadEngine extends Engine {
  prepareNext(item: PlaybackItem): Promise<StreamInfo | null>;
  advance(expected: PlaybackItem | null): Promise<StreamInfo | null>;
}

/** Result of a taken crossfade transition. */
export interface CrossfadeResult {
  /** The incoming track's stream facts (as from open()/advance()). */
  info: StreamInfo;
  /** Output-rate frames of overlap actually mixed. The caller shrinks the
   *  outgoing track's effective length by this so offsets, positions and
   *  end-of-queue detection stay truthful after the compressed boundary. */
  overlapFrames: number;
}

/** Crossfade capability (M8, opt-in). An engine implementing this can
 *  transition to `next` with an equal-power overlap instead of the normal
 *  sequential handoff. Returning null (or not implementing) makes the
 *  caller fall back to the normal EOS path — crossfade is best-effort by
 *  design. The standby/policy-agreement rule applies here too: `next` is
 *  the core's current policy target, and a standby prepared for a different
 *  item must be refused (null), never faded in. After a taken fade the
 *  engine suppresses the OLD lane's 'eos'; the new lane's lifecycle
 *  (eos/position) is the engine's responsibility, exactly as after a normal
 *  advance(). */
export interface CrossfadeEngine {
  beginCrossfade(next: PlaybackItem, fadeSeconds: number): Promise<CrossfadeResult | null>;
}

/** Pull-decode gate capability (e.g. the WASM musepack pipeline). */
export interface DecodeGate {
  start(): void;
  stop(): void;
}

export function isPreloadEngine(e: Engine): e is PreloadEngine {
  return e.capabilities.preloadNext && 'prepareNext' in e && 'advance' in e;
}

export function decodeGateOf(e: Engine): DecodeGate | null {
  if (!e.capabilities.decodeGate) return null;
  const g = e as unknown as Partial<DecodeGate>;
  return typeof g.start === 'function' && typeof g.stop === 'function' ? (e as unknown as DecodeGate) : null;
}
